### PR TITLE
Adding haptics context

### PR DESCRIPTION
**Pull request by ZoroarkZark**

### CODE REVIEW

File: lib/HapticsManager.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { atom, useAtom, useSetAtom } from "jotai"
+import { HapticEvent, HapticsFunctions, expoPlayHaptics } from "./Haptics"
+
+export const hapticsAtom = atom(true)
+
+export class HapticsManager implements HapticsFunctions {
+  play = async (hapticTrigger: HapticEvent) => {
```
Comment (ZoroarkZark) [inline lib/HapticsManager.tsx:7]: No need for async

File: lib/HapticsManager.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    const [areHapticsOn, setAreHapticsOn] = useAtom(hapticsAtom)
```
Comment (ZoroarkZark) [inline lib/HapticsManager.tsx:8]: Can't use atom in class, use atomWithStorage and set asyncstorage value in class instead

File: lib/HapticsManager.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    if (areHapticsOn) {
+      return await expoPlayHaptics(hapticTrigger)
```
Comment (ZoroarkZark) [inline lib/HapticsManager.tsx:10]: Can be a function in the class itself, instead of exported from somewhere else

File: lib/HapticsManager.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    setHapticsOn(true)
+  }
+}
+function playAction () {
+  throw new Error("Function not implemented.")
+}
+
+function errorHandle () {
+  throw new Error("Function not implemented.")
+}
```
Comment (ZoroarkZark) [inline lib/HapticsManager.tsx:32]: Get rid of these placeholder functions

File: lib/HapticsManager.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  mute = () => {
+    const setHapticsOn = useSetAtom(hapticsAtom)
+    setHapticsOn(false)
+  }
+
+  unmute = () => {
+    const setHapticsOn = useSetAtom(hapticsAtom)
+    setHapticsOn(true)
+  }
```
Comment (ZoroarkZark) [inline lib/HapticsManager.tsx:24]: Similar to earlier useAtom feedback, can't use atoms in class

File: lib/Haptics.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import * as ExpoHaptics from "expo-haptics"
+import { ReactNode, createContext, useContext } from "react"
+import { HapticsManager } from "./HapticsManager"
+
+/**
+ * A haptic event to be played to the user.
+ */
+export interface HapticsFunctions {
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:8]: Could append tiF to interface

File: lib/Haptics.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  unmute: () => void
+}
+
+/**
+ * A function type to indicate haptic feedback being played to the user.
+ */
+export type HapticEvent =
+  | "light tap"
+  | "medium tap"
+  | "heavy tap"
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:20]: Mainly using just "light tap" for now

File: lib/Haptics.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+const HapticsManagerContext = createContext<HapticsManager | undefined>(
+  undefined
+)
+
+export const useHapticsFunctions = () => useContext(HapticsManagerContext)!
+
+export type HapticsProviderProps = HapticsManager & {
+  children: ReactNode
+}
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:34]: Instead of adding in HapticsManager, add a haptics property with the interface used here

File: lib/Haptics.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
+import * as ExpoHaptics from "expo-haptics"
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:1]: Could move the contents of this file into Haptics.tsx, larger file but more organised

File: tests/components/HexColorPicker.test.tsx [hunk 10]
```diff
@@ -24,35 +27,36 @@ describe("HexColorPicker tests", () => {
...
+    expect(startWithEvent).toHaveBeenCalled()
   })
 
-  it("plays haptics on iOS when the color selection changes", () => {
-    setPlatform("ios")
+  it("calls turnOffHaptics correctly and switches the current value to what it should be", async () => {
     renderHexColorPicker(testOptions)
     selectColor(testOptions[1].color)
-    expect(hapticsPlayer).toHaveBeenCalledWith(HapticEvent.SelectionChanged)
+    expect(hapticsAtom.read).toEqual(true)
```
Comment (ZoroarkZark) [inline tests/components/HexColorPicker.test.tsx:39]: Want to test the behavior of an atom, not whether or not the atom has a value

File: lib/Haptics.tsx [hunk 11]
```diff
@@ -1,52 +1,77 @@
...
+      expoPlayHaptics(hapticsTrigger)
+    }
+  }
+
+  async mute () {
+    await AsyncStorage.setItem("haptics on/off", "false")
+  }
+
+  async unmute () {
+    await AsyncStorage.setItem("haptics on/off", "true")
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:25]: "haptics on/off" -> "@haptics_is_muted"

File: lib/Haptics.tsx [hunk 12]
```diff
@@ -1,52 +1,77 @@
...
-  | "heavy tap"
-  | "selection"
-  | "success"
-  | "warning"
-  | "error"
+  | { hapticsInteraction: "light tap" }
+  | { hapticsInteraction: "selection" }
+  | { hapticsInteraction: "success" }
+  | { hapticsInteraction: "warning" }
+  | { hapticsInteraction: "error" }
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:43]: Only need selection for now

File: lib/Haptics.tsx [hunk 13]
```diff
@@ -1,52 +1,77 @@
...
 
-const HapticsManagerContext = createContext<HapticsManager | undefined>(
-  undefined
-)
+const ExpoHapticsImplementationContext = createContext<
+  ExpoHapticsImplementation | undefined
+>(undefined)
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:47]: Don't call expo, since we could switch it to any other type of haptics

File: lib/Haptics.tsx [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+  haptics,
+  ...props
+}: HapticsProviderProps) => (
+  <ExpoHapticsImplementationContext.Provider value={haptics}>
+    {children}
+  </ExpoHapticsImplementationContext.Provider>
+)
+
+/**
+ * Haptics Player implemented by Expo.
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:71]: Documentation needed

File: lib/Haptics.tsx [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+export const useHapticsFunctions = () =>
+  useContext(ExpoHapticsImplementationContext)!
+
+export type HapticsProviderProps = {
+  children: ReactNode
+  haptics: ExpoHapticsImplementation
+}
+/**
+ * Provider for ExpoHapticsImplementation.
+ *
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:58]: Documentation needed

File: lib/Haptics.tsx [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+  async unmute () {
+    await AsyncStorage.setItem("haptics on/off", "true")
+  }
+}
+
+export interface HapticsFunctions {
+  play(hapticsTrigger: HapticEvent): Promise<void>
+  mute(): Promise<void>
+  unmute(): Promise<void>
+}
```
Comment (ZoroarkZark) [inline lib/Haptics.tsx:33]: Documentation needed

**Review by mhayes853 — APPROVED** [review]

**Pull request merged**